include/boost/serialization/singleton.hpp

```diff
@@ -150,16 +192,12 @@ class singleton {
         return get_instance();
     }
     BOOST_DLLEXPORT static bool is_destroyed(){
-        return get_is_destroyed();
-    }
-    BOOST_DLLEXPORT singleton(){
-        get_is_destroyed() = false;
-    }
-    BOOST_DLLEXPORT ~singleton() {
-        get_is_destroyed() = true;
+        return detail::singleton_wrapper< T >::is_destroyed();
     }
 };
 
+// Assigning the instance reference to a static member forces initialization
+// at startup time as described in http://tinyurl.com/ljdp8
 template<class T>
 T & singleton< T >::m_instance = singleton< T >::get_instance();
```

Comment:
Why do this at all?  On the first call to get_instance() for a type, it would instantiate the class when needed.

Reply:
Because it is needed to be instantiated at startup time (requirement of Boost.Serialization)

This is checked by the `locked` feature https://github.com/boostorg/serialization/pull/131/files#diff-3f1e861be15fd4926f5e5b00388f2309R188

Reply:
I see, the requirements are to avoid locking so this needs to be done at initialization time.  This is a pretty funky way to do it. :)

Reply:
That's why I added the comment and a reference so future readers know why this is done. I deem this an important part of this PR

Reply:
Note on locking.  Strictly speaking this "locking" feature is not necessary.  In the serialization library I wanted all my singletons to be constant - ie not mutable.  I wanted to trap attempts to modify them after main is called.  But of course that means I want to force creation on program/shared library load.  So I'm fine with the comment.  I would recommend that the comment not point to a link - even if the text is longer.  The serialization library will likely persist in usage long after the link is gone.

Reply:
I actually think it is required as a test that construction happens before main. If it would be constructed on first use, you won't be able to deserialize a derived class from a pointer for which you didn't "use" the singleton yet (e.g. by serializing it or refering to it otherwise)

I can remove the link if you wish or replace it by the original: https://groups.google.com/forum/#!topic/microsoft.public.vc.language/kDVNLnIsfZk (in the hope that even in the unlikely case google will vanish services as archive.org can be used to retrieve it), but the essentials are summarized in the 1 sentence already. Your call.